Propose NuGet.org portions of TFM spec (#10573)
NuGet/Home: proposed/2021/NuGet.orgTFMs.md

Q: Will the "Frameworks" section be expanded by default? Should the "Frameworks" section explain how the customer should use this information? Should we add a link to docs? Consider moving up the "Frameworks" section. I would suggest "Documentation" first, followed by "Frameworks". The "Version History" table is noisy and it can be difficult to see what's below it. The "Dependencies" and "Used By" sections are useful information, but would claim "Frameworks" is more important.
A: Let's not expand by default--this will give us the option of trackiong the popularity of the section by tracking expansion clicks.

Q: Potentials for modelling our UX on:
 - https://docs.microsoft.com/en-us/dotnet/api/system.text.stringbuilder?view=net-5.0#applies-to
 - https://apisof.net/catalog/e40ef6ad-bd6e-9f29-7864-b942e37be11f
 - https://docs.microsoft.com/en-us/dotnet/standard/frameworks#latest-versions
A: From a sample of 100,000 package/versions, we have:
 - Only 445 have more than 5 supported TFMs
 - Only 16 have more than 10
 - Only 1 more than 15 (and it has 16)

>  So, we'll go with a flat list for now and look at other options for subsequent work if desired.

Reviewer: I can also work with folks on the UX side, accessibility of information here is more of a PM/UX question so happy to chase that down

## Motivation

### Problem: What problem is this solving? 

Developers cannot use the package details page to determine what frameworks a NuGet package can support. In 2021, we live in a world where there are many evolutions of .NET coexisting on NuGet, such as .NET Framework and .NET 5.

Developers should be able to determine whether a package works for their project without having to first download it or rely on the author’s documentation. 

### Why: How do we know this is a real problem and worth solving? 

Finding a package that is compatible with your project is a challenge for the average developer in the .NET ecosystem. Over 21% of developers fail to install a package today. One of the top NuGet errors issued in telemetry is [NU1202](https://docs.microsoft.com/en-us/nuget/reference/errors-and-warnings/nu1202), which shows that we allow developers to install packages that are not compatible with their project or solution’s target framework. 

### Success: How do we know if we’ve solved this problem? 

- By gauging excitement/disappointment after blogging about this feature & analyzing sentiment on Twitter/GitHub/DevCom/etc. 

> 

Reviewer: - Compare the before and after numbers for error messages due to TFM incompatibility. Is there a noticeable decrease? If so, how long after the feature was rolled out for the decrease?

Not a perfect metric, but some data to show

Reviewer: Based on the convo below, how many people click open the section to expand.

Author: Added this metric to the spec.

### Audience: Who are we building for? 

We are building this for the .NET developer who is browsing & making decisions about including a NuGet package into their project based on target framework & supported platform criteria. 

## Explanation

### Functional Explanation

Context and Scope: When a developer accesses the details page for a NuGet package, they will see information on target frameworks in an expandable section. The framework information will be surfaced from the package.

Minimal Requirements:
- List of Target Frameworks on NuGet.org's package details page.
- Runtime, Content Build assets and Tools packages will display TFM details.
- Support for PackageReference project restore compatibility only.

**Target Framework** – A target framework is a specification of the APIs available to an application or library. 

Example: .NET Framework, .NET Core, .NET Standard, .NET 5 are target frameworks. 

**Supported Platform** – A supported platform is a target that the APIs can be ran on. 

Example: iOS, Android, Windows are platforms. 

- A .NET developer can view Frameworks of a package on NuGet.org: 
![](../../meta/resources/NuGet.orgTFMs/PackageDetailsWithTFMs.png)

> - We will have the list collapsed by default in order to track popularity.

Reviewer 2: At an earlier point in time (about ~1 year ago), we did some analysis on `what frameworks does a package really support` vs `what frameworks are declared in the dependencies section` and we found that for about 95% of packages, these are a 1 to 1 relationship. 

Have we consider making that more apparent instead? In the interest of saving some vertical space/introducing a new concept that's already covered in most scenarios.

Reviewer 3: This is a good point. There's a challenge here where you run into the XUnit case.

![image](https://user-images.githubusercontent.com/1288848/108431455-81ec8b80-7208-11eb-8b52-d6ab4e8b737b.png)

To get to an actual "idea" of if the package supports a specific framework is layers deep to the dependencies defined such as:

![image](https://user-images.githubusercontent.com/1288848/108431520-9597f200-7208-11eb-8c86-c5b520fb7522.png)

So in these cases, we don't have a good answer without iterating dependencies.

Reviewer 2: That totally makes sense. 

I just thing integrating it differently, not in an additive way might save us some space. Right now there's *a lot*.

Reviewer 3: We just need to consider a tab design for these pages as a "Future Possibility". `Dependencies`, `Frameworks`, `Version History`, `Dependents (Used By)`, etc really belong in their own individual tabs so everything isn't stacked ontop of each other. `README` for example will make scrolling to get to one of these a default experience.

/cc @jcjiang may be worth adding to this section!

Author: Discussed offline and we'll consider these to be universally compatible at this stage. I've added `- A design for meta packages` to future possibilities, and splitting expanders out to tabs may be called out here or in anothr spec. Up to you, @jcjiang. As we've nailed down the pertinent corner cases, for now perhaps we can merge this?

# NuGet Target Frameworks on Package Details Page 

- Author Name (https://github.com/jcjiang)
- Start Date (2021-02-16)

> - GitHub Issue (https://github.com/NuGet/NuGetGallery/issues/8387)

Reviewer 4: There are a bajillion issues about package compatibility. Have we scoured them for clues on how to improve this limited-scope proposal? Many asks are about searching or VS UI which are out of scope, but I imagine some feedback can be considered right now.

Reviewer 4: E.g. https://github.com/NuGet/NuGetGallery/issues/3098

Reviewer 4: This issue have many links to and from it.

Reviewer 3: I believe https://github.com/NuGet/NuGetGallery/issues/4843 is the best issue to describe this scope.

Author: I love how these 4-year-old conversations just get peeled off in these discussions like they were sitting at top of mind for the encyclopedic amongst us. :-D

Author: I've moved @terrajobst's issue front and center in this spec, and will close the tracking issue I've been using and use Immo's.

- GitHub PR ()

## Summary

> When a developer opens the package details page or tab on NuGet.org or NuGet rooling, they will see what frameworks the package can support. 

Reviewer 5: ```suggestion
When a developer opens the package details page or tab on NuGet.org or NuGet tooling, they will see what frameworks the package can support. 
```

Author: I think tooling shouldn't be here. I stripped out most of them and missed this one.

Author: Removed.

## Motivation

### Problem: What problem is this solving? 

> Developers cannot use the package details page to determine what frameworks a NuGet package can support. In 2021, we live in a world where there are three evolutions of .NET coexisting on NuGet; .NET Framework, .NET Core, and .NET 5.  

Reviewer 5: There's also Mono, Unity, WebAssembly, NativeAOT. All of these runtimes have non-trivial usage and significant implications. I'd consider wording this in a more open-ended way:

```suggestion
Developers cannot use the package details page to determine what frameworks a NuGet package can support. In 2021, we live in a world where there are many evolutions of .NET coexisting on NuGet, such as .NET Framework and .NET 5.
```

Author: I'll defer to @jcjiang on this.

Reviewer: The change makes sense, always good to be more open-ended and accurate.

Developers should be able to determine whether a package works for their project without having to first download it or rely on the author’s documentation. 

### Why: How do we know this is a real problem and worth solving? 

Finding a package that is compatible with your project is a challenge for the average developer in the .NET ecosystem. Over 21% of developers fail to install a package today. One of the top NuGet errors issued in telemetry is [NU1202](https://docs.microsoft.com/en-us/nuget/reference/errors-and-warnings/nu1202), which shows that we allow developers to install packages that are not compatible with their project or solution’s target framework. 

### Success: How do we know if we’ve solved this problem? 

> - By measuring the change in the percentage of successful installs and lowering the amount of [NU1202](https://docs.microsoft.com/en-us/nuget/reference/errors-and-warnings/nu1202) errors in telemetry. 

Reviewer 4: Does this make sense prior to the filtering work? The information will not be surfaced in VS in this phase so I imagine the VS behavior won't change significantly (not sure though)

Reviewer 3: I think this is largely an artifact of the old spec I wrote. @jcjiang or @drewgillies should be able to clean it up to outline the scope of the details page only for nuget.org.

Author: Removed.

> - By measuring the change in the percentage of successful installs and lowering the amount of [NU1202](https://docs.microsoft.com/en-us/nuget/reference/errors-and-warnings/nu1202) errors in telemetry. 

Reviewer 4: If the section is collapsed by default on the UI, we can measure how often users expand the section. We also measure how often the copy command button is clicked. We could consider observing these metrics to gauge success. Just an idea.

Reviewer 3: Another thought is % of users opening up dependencies vs. frameworks & doing the same you suggest.

Author: I've added this.

- By measuring the usage of target framework & supported platform filters on NuGet.org & Visual Studio. 
- By gauging excitement/disappointment after blogging about this feature & analyzing sentiment on Twitter/GitHub/DevCom/etc. 

### Audience: Who are we building for? 

We are building this for the .NET developer who is browsing & making decisions about including a NuGet package into their project based on target framework & supported platform criteria. 

## Explanation

### Functional Explanation

> #### Context and Scope 

Reviewer 3: I think this header should be removed & contents moved into the spec. The context & scope is already defined in the "Summary" & "Motivation" sections.

Author: Done.

> When a developer accesses the details page for a NuGet package, they will see information on target frameworks in a dropdown menu. The framework information will be surfaced from the package. In cases where versions are not specified (such as .net) or other unexpected scenarios, ‘Version Not Specified’ or similar messages will be surfaced. 

Reviewer 5: I'd consider changing `dropdown menu` to `expandable section` as dropdown menu has a slightly different connotation. For example: https://www.w3schools.com/howto/howto_css_dropdown.asp

```suggestion
When a developer accesses the details page for a NuGet package, they will see information on target frameworks in a expandable section. The framework information will be surfaced from the package. In cases where versions are not specified (such as .net) or other unexpected scenarios, ‘Version Not Specified’ or similar messages will be surfaced. 
```

Reviewer 3: > In cases where versions are not specified (such as .net) or other unexpected scenarios, ‘Version Not Specified’ or similar messages will be surfaced.

These would be considered v0.0 and should show in the same way.

For unexpected scenarios, we should not show anything.

Author: Both changed.

> When a developer accesses the details page for a NuGet package, they will see information on target frameworks in a dropdown menu. The framework information will be surfaced from the package. In cases where versions are not specified (such as .net) or other unexpected scenarios, ‘Version Not Specified’ or similar messages will be surfaced. 

Reviewer 5: > In cases where versions are not specified (such as .net) or other unexpected scenarios, ‘Version Not Specified’ or similar messages will be surfaced. 

I'm guessing this is referring to target frameworks without a version? Should this be moved down to the `Technical Explanation` section? This is very a specific concern for the `Context and scope` section

Author: Moved.

> ##### Minimal Requirements 

Reviewer 3: Requirements header can be dropped & consolidated into the functional design. Only include the minimal requirements, and not the non-goals.

Author: Done.

> - List of Frameworks on NuGet.org

Reviewer 4: - What to do with non-Dependency type packages? e.g. .NET tools?
- What about `tools` assets in dependency packages? will `tools/net40` make the package appear as compatible with net40?

Author: _Do_ we consider `tools/net40` to be compatible with `net40`?

Reviewer 3: Yep. `tools/{TFM}` would represent that the tool is supported on that version of .NET. We should list it the same way in the sense that a tool can be compatible with net5.0 / netcoreapp3.1 / etc.

Author: Added `- Runtime, Content Build assets and Tools packages will display TFM details.`

Reviewer 4: `tools/{tfm}/{rid}/{any}` is not used for dependency package compatibility ([source](https://github.com/NuGet/NuGet.Client/blob/97311b5cb48d38402b49fd41a6384b11db7dd453/src/NuGet.Core/NuGet.Commands/RestoreCommand/CompatibilityChecker.cs#L431-L443)). If we choose to display it, people theoretically could still encounter failed installs with based on this information.

Author: Ah, that's interesting. Yes, that's the risk of including a pattern set that's not called out in the restore code's compat checks. This was a tenuous inclusion anyway, and I can remove it if we don't want this kind of misleading. Build assets could give us the same problem though, right?

Reviewer 4: Build assets are considered in the compat check:
https://github.com/NuGet/NuGet.Client/blob/97311b5cb48d38402b49fd41a6384b11db7dd453/src/NuGet.Core/NuGet.Commands/RestoreCommand/CompatibilityChecker.cs#L441-L442

Author: Not the same set, by the looks: https://github.com/NuGet/NuGetGallery/blob/dg-tfmbackfill/src/GalleryTools/Commands/BackfillTfmMetadataCommand.cs#L65-L67
For a purer experience, perhaps I should remove the transitive build pattern set.

Returning to tools, if we present a TFM based on the tools pattern set, and there is no compat checking on restore, isn't it more likely that the _NuGet.org_ experience is more restrictive? In other words, would we see TFMs on some tools packages that would always restore, but not all restorable tools packages would have TFMs on NuGet.org?

Author: After discussion offline, we're looking to:
- keep build asset patterns but drop out transitive build assets in order to make this compat check consistent with nuget restore
- only assess tools pattern sets in the case of tools packages (also consistent with restore logic).

This is beyond the scope of a UX spec, of course, but relevant in making it happen.

> - List of Frameworks on NuGet.org

Reviewer 4: I don't know if there are any differences between PackageReference and packages.config compatibility. Client team will know for sure. If there is, we should note that this is _PackageReference_ compatibiliy

Reviewer 3: @nkolev92 or @zivkan have any thoughts here?

Reviewer 6: There are some differences, yes.

For example, packages.config will accept a package with `lib/foo.dll`, and this package will be deemed compatible with all TFMs. But PackageReference will completely ignore it, and only accept packages with `lib\<tfm>\foo.dll` (assuming the package doesn't have other assets.

I haven't yet read this spec, so I don't know if my comment makes sense in whatever context this section of the document is, but it's absolutely possible for a package to work only for packages.config, only for PackageReference, or work for both.

Reviewer 4: Your comment makes sense in whatever context this is 🥇

Author: This is pertinent to our TFM-assessment logic on package push. If we're looking to support `PackageReference`, we need to not fall back to nuspec reads when the package doesn't contain `lib\<tfm>\foo.dll`, and there's definitely no case for inspecting dlls for target framework metadata.

Author: Added `- Support for PackageReference project restore compatibility only.`

> - List of Frameworks on NuGet.org

Reviewer 5: Community members that read this may be sad that this doesn't mention Visual Studio. Are we planning on adding Visual Studio support in the future? Do we have a GitHub issue we can link to so that they can get updates on that?

Reviewer 3: This should be listed in the "Future Possibilities" section.

Author: Added a Future Possibilities section with the following 
```
- Filtering NuGet.org search by TFM
- TM display in Visual Studio
```

> ##### Non-Requirements 

Reviewer 3: This is redundant & not in the scope of this proposal. We don't need non-requirements.

Author: Removed.

> - List of Platforms on NuGet.org

Reviewer 5: I'd consider defining what platform means for folks that aren't on the NuGet team.

  - Users interviewed did not express need for supported platform information 
  - Cannot be surfaced directly from package data and would require additional layer of soft calculation from engineering (no clear mapping scheme currently exists between target frameworks and supported platforms) 
  - Can be revisited in future 
- Filters for Frameworks on NuGet.org

>     - Users interviewed were ambivalent on filters, requires further investigation. Can be revisited in future 

Reviewer 5: Does this mean users were ambivalent on searching by a specific target framework? This issue seems to indicate customers feel pretty strongly about filtering search results that are incompatible with my project: https://github.com/NuGet/NuGetGallery/issues/3098

Reviewer 3: Frameworks are really platforms pre .NET 5. .NET 5+ we can leverage platforms more. 

I think there's a hidden requirement here in which the "platform is implied by the TFM". At least from many people I've talked to, a TFM is synonymous with a platform today but will change in NET 6 drastically.

Reviewer 3: In my eyes we do the following:

TFM -> Platforms -> Runtimes

Author: I believe the question was about the priority of filtering, and I'll call this good for this iteration. My understanding is that this will be spec'd and implemented separately from the Package Details UX.

- Filters for Platforms on NuGet.org
  - Users interviewed were ambivalent on filters, requires further investigation. Can be revisited in future 

Although Target Frameworks & Supported Platforms are relatively similar in general contexts and often have overlap, we must clearly define each of these items based on their differences. Today we live in an ecosystem where Target Frameworks can represent both a supported platform, or a subset of supported platforms. 

**Target Framework** – A target framework is a specification of the APIs available to an application or library. 

Example: .NET Framework, .NET Core, .NET Standard, .NET 5 are target frameworks. 

**Supported Platform** – A supported platform is a target that the APIs can be ran on. 

Example: iOS, Android, Windows are platforms. 

> ### Solution 

Reviewer 3: The solution should be included in the "Functional Design" section. No extra headers are needed!

Author: An artifact from the doc spec. Gone.

- A .NET developer can view Frameworks of a package on NuGet.org: 

> ![](../../meta/resources/NuGet.orgTFMs/PackageDetailsWithTFMs.png)

Reviewer 4: Expanded by default? (dependencies are not)
Why isn't this the first section, even above dependencies?

Reviewer 4: What order are the frameworks in? Lex order? Or some order we predefine?

Reviewer 3: framework precedence would be a great ordering.

Author: How do we define framework precedence?

Reviewer 3: https://nugettoolsdev.azurewebsites.net/5.8.0/framework-precedence 

There's a few APIs in NuGet.Frameworks IIRC.

Reviewer 6: I think that framework precedence depends on project target framework. Or at least, with the current discussions about iOS/Mac TFMs in net6.0 it will be.

Author: Thanks @zivkan. So there may be a future-proofing concern here.
Just checking that we're not getting away from the intent of the spec a bit on this one?

Reviewer 3: Naw, just that there's mostly a natural ordering we can still follow based on the net5.0 precedence defined today or shown similarly in the latest target framework docs!

Author: I like the notion of not expanded by default to allow for tracking as you mentioned elsewhere. I've added this, but it can always be removed if this isn't where we choose to land: `- We will have the list collapsed by default in order to track popularity.` 
I'd like it before Dependencies also, and will defer to @jcjiang on this.

Author: > Naw, just that there's mostly a natural ordering we can still follow based on the net5.0 precedence defined today or shown similarly in the latest target framework docs!

Fair enough. If there are APIs it should be easy enough to implement, and more importantly, it won't need maintenance.

Reviewer: I also would like it before Dependencies - let's make the change. T

otally agree on not expanded by default to allow for tracking, could be another metric we can track for success. Are people even clicking open the section? And based on that, we could then toggle for automatic showing if that makes it easier for people to find the information. But I don't want to start off with automatic showing and not be sure if it's useful.

Author: Great--included this change.

> ![](../../meta/resources/NuGet.orgTFMs/PackageDetailsWithTFMs.png)

Reviewer 4: Consider a display mimicking the API docus:
https://docs.microsoft.com/en-us/dotnet/api/system.text.stringbuilder?view=net-5.0#applies-to
There's also Immo's https://apisof.net/catalog/e40ef6ad-bd6e-9f29-7864-b942e37be11f

Reviewer 3: Looks great. Going back to framework precedence order it by "Latest TFM -> Oldest least compatible TFM"

Reviewer 3: https://docs.microsoft.com/en-us/dotnet/standard/frameworks#latest-versions in other words

Author: Are we leaving our intended MVP scope behind a little? This would be impressive and might be a good v2 feature, but I'm concerned that we're throwing some cost in here where we don't need to. Can we backlog this one?

Reviewer 3: I don't think there's extra cost here. Just deciding on whether a format like the existing `Dependencies` section or a table like `Version History` would represent this information better given NuGet.org is already quite heavy on vertical real estate.

Tables have pros in terms of saving vertical space at the cost of horizontal space.

Headers have pros in terms of saving horizontal space at the cost of vertical space.

/cc @jcjiang

Author: Let's step back a moment. This seems predicated on us not just listing actual TFMs present in the archive (as originally floated), but on exploding this to every possible supported TFM. I want to be sure that we're intentionally adding this to the spec, because there's cost there. That's my cost concern--not &lt;br /&gt; vs. &lt;tr /&gt; tags.

Author: If this is to be added to the spec, we'd need to call it out explicitly, and I'd suggest either:
- put it in as a nice-to-have, not an MVP item, or
- backlog it for the future

Reviewer 3: If we put the technical cost aside since a user could multitarget the world & have every single TFM in their archive

i.e. Microsoft.CSharp

![image](https://user-images.githubusercontent.com/1288848/108278866-4dfa6300-7141-11eb-9f10-b0dbbe1fb3ad.png)

This gets rather unwieldy quickly in terms of vertical space.

Reviewer 3: This example wouldn't help much with a table as only the `netstandard` items would likely save space. This one is just unique 😂

Reviewer 3: Should just resolve this feedback & move on with the original mockup if it's easiest! Maybe put under a `Unresolved Questions` if @jcjiang wanted to dive more in the UX w/ the UX team here.

Author: Yep, that works for me if @jcjiang is good with that approach. Good example! It's always good to look at nth degree cases like this.

Reviewer: Agreed, let's put this under unresolved questions and I will schedule some time with the UX team when the time comes.

Author: I've added this as an unresolved question.

> ![](../../meta/resources/NuGet.orgTFMs/PackageDetailsWithTFMs.png)

Reviewer 4: For new users, these may be rather opaque strings. Can we link to a "good" doc that the .NET team has (don't know which one) that explains these various frameworks.

Said, if I install the .NET SDK or Visual Studio and I want to hack away at a project, what good first document can I read to start understand this wonderful ✨ world of TFMs we have.

Reviewer 4: I'm sorry, I now see `Unresolved questions`. This covers some of my questions.

Reviewer 3: https://docs.microsoft.com/en-us/dotnet/standard/frameworks#latest-versions imo

Author: Great--I've added this to the Target Framework List section:
```
Let's link to an explanatory document for any frameworks we expose here, for example: https://docs.microsoft.com/en-us/dotnet/standard/frameworks#latest-versions
```
@jcjiang--I'm not entirely sure how this would need to look in UI. I initially asked you for a mockup, but I think it would just be the first text we see when expanding. Might need your help on wording though, and anything else we might want in there.

Reviewer: Makes sense. Ping me when you are messing with wording, happy to help there.

### Goals and Non-Goals 

#### Goals 

> Make it clear to developers whether a package can be installed or updated for their project or solution based on a target framework. 

Reviewer 4: We should be clear whether we are listing all frameworks that are supported _or_ exact frameworks that have assets.

Consider a simple package with `lib/netstandard1.0/foo.dll`. Are we going to just say `netstandard1.0` or are we going to list all of the frameworks that support `netstandard1.0`.

Reviewer 3: just netstandard1.0. we don't care about the whole list b/c that can be looked up by users. We should link to tables for any portable assets though like https://docs.microsoft.com/en-us/dotnet/standard/net-standard#net-implementation-support

Author: The logistics concern me there--we would need to maintain a dictionary, which could mean that we have the potential for it to contain errors or fall out of maintenance. Those who use abstracted APIs like netstandard would usually understand the abstraction, but we could provide a link to a conversion matrix a la https://docs.microsoft.com/en-us/dotnet/standard/net-standard, which given its location in docs should be kept up to date.

Reviewer 3: There's the other consideration that portable assets != platform assets & therefore we can't just list supported platforms as we can mislead users.

Author: That's an important distinction. I've added in the Technical Explanation section:
```
- When a portable TFM is being broken up and displayed (if feasible), we will need to flag it as portable, as portable assets are not the same as platform assets.
```

#### Non-Goals 

> Over-complicate what it means to be a compatible package. One target framework + one target platform = compatible package. 

Reviewer 4: This is spicy, lol 🤣

Reviewer 3: lol even spicier is that I believe you defined this early on 🌶️

Reviewer 4: oh RIP ⚰️

Reviewer 3: f in chat.

Reviewer 5: f

Author: Okay, this really doesn't need to be here, does it. I'll remove it. :)

### Design 

#### Target Framework List 

- .NET Framework (net) 
- Xamarin/Mono (monoandroid, monotouch, xamarinios, monomac, xamarinmac, xamarinwatchos, xamarintvos) 
- .NET Core (netcoreapp) 
- .NET Standard (netstandard) 
- .NET 5 (net5.0) 

### Technical Explanation

Packages whose layout follow our [guidelines](https://docs.microsoft.com/en-us/nuget/create-packages/supporting-multiple-target-frameworks) will be examined when uploaded and validated, prior to listing. Target Frameworks will be determined and displayed here.

Note that packages which do not follow the guidelines may have inconsistent results.

> ### Unresolved questions

Reviewer 4: Could we have a general Q&A/FAQ section where these open questions get moved to Q & A when they are resolved? Perhaps there are common questions that the team or community will have about this spec that could be documented.

Author: We could move it to https://github.com/NuGet/Home/discussions/new ...

Reviewer 3: I think there's a few missing sections from the new template that would help out here further such as what was considered & what alternatives exist so there's less FAQ/QnA loops. Also the unresolved questions section should keep a tallied list of what comes up in PR reviews like this!

Reviewer 3: https://github.com/NuGet/Home/blob/dev/meta/template.md if you're interested!

Author: I made this a hybrid between the template and the doc spec, but I agree--it would be good to add those sections back in.

Author: I've added all but Prior Art back in.

> ### Unresolved questions

Reviewer 6: I think `build/<packageid>.[props|targets]` poses a problem, because from NuGet's point of view this package is compatible with all TFMs, .NET, native/c++, or anything else. For packages like Nerdbank.GitVersioning, or SourceLink, this makes sense.  But many .NET packages incorrectly use this, rather than `build\<tfm>\<packageid>.[props|targets]`. I think it's worthwhile pointing this out and explaining what our plan is for these packages.

Author: I've moved this into unresolved questions, @zivkan.
